### PR TITLE
Start Adding .NET 4.6 APIs

### PR DESCRIPTION
**Pull request by borgdylan**
Adding Array.Empty<T> + Tests
See reference source PR: https://github.com/mono/referencesource/pull/13 for Tasks related .NET 4.6 stuff
Adding xbuild-frameworks for .NET 4.5.2 and .NET 4.6

Signed-off-by: Dylan Borg <borgdylan@hotmail.com>

### CODE REVIEW

**Comment by tritao** [comment]
approve

**Comment by borgdylan** [comment]
Current signing CLA. The tests I added pass on my machine.

**Comment by borgdylan** [comment]
Also, more PRs will be on the way once this is merged.

**Comment by borgdylan** [comment]
@tritao a related pr is here https://github.com/mono/referencesource/pull/13

**Comment by borgdylan** [comment]
The failure reported for ARM soft float does not seem to originate from my PR.

**Comment by kumpera** [comment]
@marek-safar can we add 4.6 method unguarded?

**Comment by borgdylan** [comment]
By guarded you mean hidden behind a define?

**Comment by borgdylan** [comment]
@kumpera I will be PR-ing most 4.6 stuff that I can manage to implement.

**Comment by kumpera** [comment]
@borgdylan exactly, not sure if we need it behind a define.

**Comment by marek-safar** [comment]
Yes, we need it under define. I assume we get whole .net 4.6 api via next reference sources upload and it'd be much easier to remove your code if it's clearly marked to avoid duplicate implementation.

**Comment by borgdylan** [comment]
@marek-safar so I need the define only for reference sources changes?

**Comment by marek-safar** [comment]
For both, we'll probably still need to generate binary-reference for 4.5

**Comment by borgdylan** [comment]
What should be name of the define? NET_4_6?

**Comment by tritao** [comment]
I think that's OK

**Comment by borgdylan** [comment]
I will update the PR. I will be enabling the define by default while building.

**Comment by borgdylan** [comment]
@marek-safar this PR is also accompanied by this one https://github.com/mono/referencesource/pull/13. I will be able to add the tests for it until it gets merged in.

**Comment by borgdylan** [comment]
The i386 test fail is due to something in System.Xaml making mono segfault. It has nothing to do with what I changed.

**Comment by tritao** [comment]
I've merged this reference sources PR, can you now also update the submodule ref on this PR?

**Comment by borgdylan** [comment]
@tritao yes. Do I need to do another PR with tests for the reference source stuff? I have not written any new complicated logic in reference source PR.

**Comment by tritao** [comment]
Yeah, we should probably add some tests covering the new APIs.

**Comment by borgdylan** [comment]
Should the tests be in this PR or in another one?

**Comment by tritao** [comment]
Either way is fine, I leave it to your preference.

**Comment by borgdylan** [comment]
Ok I will add them to this one so its all a cohesive whole. I will run them on my machine first.

File: mcs/class/corlib/Makefile [hunk 1]
```diff
@@ -43,12 +43,12 @@ ifndef MOBILE_STATIC
 REFERENCE_SOURCES_FLAGS += -d:FEATURE_REMOTING,MONO_COM,FEATURE_COMINTEROP,FEATURE_ROLE_BASED_SECURITY
 endif
 
-LOCAL_MCS_FLAGS = -unsafe -nostdlib -nowarn:612,618 -d:INSIDE_CORLIB,MONO_CULTURE_DATA -d:LIBC $(REFERENCE_SOURCES_FLAGS)
+LOCAL_MCS_FLAGS = -unsafe -nostdlib -nowarn:612,618 -d:INSIDE_CORLIB,MONO_CULTURE_DATA,NET_4_6 -d:LIBC $(REFERENCE_SOURCES_FLAGS)
```
Comment (akoeplinger) [inline mcs/class/corlib/Makefile]: This isn't the correct way of setting the define, you need to create a new profile similar to [net_4_5.make](https://github.com/mono/mono/blob/master/mcs/build/profiles/net_4_5.make) and plumb it throughout `configure.ac` etc.

**Comment by akoeplinger** [comment]
I wonder whether this is a good way forward? Now that we're essentially always running on a single set of assemblies, should we just get rid of the net_4_* profile and call it net_current? I'm assuming once referencesource is updated with the 4.6 sources, we'll automatically have all the new APIs anyway unless we go back and ifdef them out for 4.5.

**Comment by borgdylan** [comment]
@akoeplinger has a decision bee reached on what I have to do to conclude the PR?

**Comment by akoeplinger** [comment]
@borgdylan my question was more directed at the Mono team, let's see what they say :)

**Comment by borgdylan** [comment]
@akoeplinger  Ok. I will change the PR according to what they say.

**Comment by borgdylan** [comment]
ping

**Comment by marek-safar** [comment]
Please make separate PR for xbuild changes.
Don't add NET_4_6 to 4.5 profile but create a new one.

**Comment by borgdylan** [comment]
@marek-safar so I will create a profile that installs to {prefix}/lib/mono/4.6 ?

**Comment by marek-safar** [comment]
Good question. I think yes we always followed this path in past

**Comment by akoeplinger** [comment]
@marek-safar as I noted above, how will this work once referencesource is updated with .NET 4.6 sources (do we need to manually add ifdefs around all the new APIs)?

I think I'd prefer getting rid of the profiles altogether and always ship a single set of assemblies for running + reference assemblies for compilation against 2.0/3.0/4.x.

**Comment by marek-safar** [comment]
@akoeplinger I agree using reference assemblies for profile targeting is preferred but we don't have it at the moment. Hence 4.5 target would actually show 4.6 api.

**Comment by borgdylan** [comment]
To avoid merging complications, can I reopen this PR after I split it in two?

**Comment by borgdylan** [comment]
WHat  will happen of this PR since 4.6 and 4.5 have been unified in mono? There is a net_4_x profile?

**Comment by borgdylan** [comment]
Since this PR is no obsolete I'd like to close it.

**Pull request closed without merging**